src/ripple/protocol/impl/STObject.cpp

```diff
@@ -696,22 +696,27 @@ bool STObject::operator== (const STObject& obj) const
 
 void STObject::add (Serializer& s, bool withSigningFields) const
 {
-    std::map<int, STBase const*> fields;
-    for (auto const& e : v_)
+    using Entry = std::pair <int, STBase const*>;
+    std::vector<Entry> entries;
+    entries.reserve (v_.size());
+
+    for (detail::STVar const& e : v_)
     {
-        // pick out the fields and sort them
+        // Keep only the fields we want.
+        SField const& sField {e->getFName()};
         if ((e->getSType() != STI_NOTPRESENT) &&
-            e->getFName().shouldInclude (withSigningFields))
+            sField.shouldInclude (withSigningFields))
         {
-            fields.insert (std::make_pair (
-                e->getFName().fieldCode, &e.get()));
+            entries.emplace_back (sField.fieldCode, &e.get());
         }
     }
 
+    std::sort (entries.begin(), entries.end());
```

Comment:
If you want to be really paranoid, you can check that there are no duplicate fieldCodes in the entries collection.

Reply:
I share both of these concerns too.  The lambda sort will also add a small bit of savings in code size (dropping a not-used compare).

Is a duplicate key a logic error or something that should be ignored?  If logic error, assert, otherwise `std::unique` could be used on the sorted vector.

Reply:
If assert, I believe this is the proper test:  `std::adjacent_find(entries.begin(), entries.end(), [](auto const& x, auto const& y) {return x.first == y.first;}) == entries.end()` (untested)